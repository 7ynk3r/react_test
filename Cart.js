import { View, Button, Text } from 'react-native';
import React from 'react';

export const Cart = ({ cart }) => (
  <Container>
    <User user={ cart.user } />
    <Items items={ cart.items } />
  </Container>
);

export const Container = (props) => <View { ...props } />

export const User = ({ user }) => (
  <Title>{ user.name + ' - ' + user.email }</Title>
);

export const Title = (props) => <Text { ...props } />

export class Items extends React.Component {
  constructor(props) {
    super(props);
    this.state = this.calculateState(props.items);
  }
  calculateState = (items) => ({
    items,
    total: items.map(item => item.price)
      .reduce((a, b) => a + b, 0),
  })
  deleteItem = ({ id }) => {
    const items = this.state.items.filter(item => item.id !== id);
    this.setState(this.calculateState(items));
  }
  render = () => (
    <Container>
      { this.state.items.map(item => (
        <Item
          key={ item.id }
          item={ item }
          onDelete={ () => this.deleteItem(item) }
        />
      ))}
      <Total total={ this.state.total }/>
    </Container>
  )
}

const Total = ({ total }) => <Text>Total { total }</Text>

const Item = ({ item, onDelete }) => (
  <Container>
    <Text>{ item.name }</Text>
    <Text>{ item.price }</Text>
    <Button title='Delete' onPress={ onDelete }/>
  </Container>
);

export default Cart;
