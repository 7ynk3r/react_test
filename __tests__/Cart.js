import { View, Button, Text } from 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

import { Cart, Items, User } from '../Cart';

describe('User', () => {
  const user = {
    name: 'joe',
    email: 'joe@mail.com',
  };

  it('renders deep correctly', () => {
    const tree = renderer.create(<User user={ user }/>);
    expect(tree).toMatchSnapshot();
  });

  it('renders shallow correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<User user={ user }/>);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });
})

describe('Cart', () => {
  const cart = {
    user: {
      name: 'joe',
      email: 'joe@mail.com',
    },
    items: [{
      id: 1, name: 'item 1', price: 7,
    }, {
      id: 2, name: 'item 2', price: 4,
    }]
  };

  it('renders deep correctly', () => {
    const tree = renderer.create(<Cart cart={ cart }/>);
    expect(tree).toMatchSnapshot();
  });

  it('renders shallow correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<Cart cart={ cart }/>);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });
});

describe('Items', () => {
  it('renders zero items correctly', () => {
    const renderer = new ShallowRenderer();
    const items = [];

    renderer.render(<Items items={ items }/>);
    const result0 = renderer.getRenderOutput();
    expect(result0).toMatchSnapshot();
  });

  it('renders one item correctly', () => {
    const renderer = new ShallowRenderer();
    const items = [{
      id: 1, name: 'item 1', price: 7,
    }];

    renderer.render(<Items items={ items }/>);
    const result0 = renderer.getRenderOutput();
    expect(result0).toMatchSnapshot();
  });

  it('renders two items and deletes correctly', () => {
    const renderer = new ShallowRenderer();
    const items = [{
      id: 1, name: 'item 1', price: 7,
    }, {
      id: 2, name: 'item 2', price: 4,
    }];

    renderer.render(<Items items={ items }/>);
    const result0 = renderer.getRenderOutput();
    expect(result0).toMatchSnapshot();

    // access the first item and call `onDelete`.
    result0.props.children[0][0].props.onDelete();
    const result1 = renderer.getRenderOutput();
    expect(result1).toMatchSnapshot();
  });
});
