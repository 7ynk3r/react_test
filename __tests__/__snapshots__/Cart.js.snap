// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders deep correctly 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
  >
    joe - joe@mail.com
  </Text>
  <View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        item 1
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        7
      </Text>
      <View
        accessibilityComponentType="button"
        accessibilityLabel={undefined}
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Delete
          </Text>
        </View>
      </View>
    </View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        item 2
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        4
      </Text>
      <View
        accessibilityComponentType="button"
        accessibilityLabel={undefined}
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Delete
          </Text>
        </View>
      </View>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      Total 
      11
    </Text>
  </View>
</View>
`;

exports[`Cart renders shallow correctly 1`] = `
<Container>
  <User
    user={
      Object {
        "email": "joe@mail.com",
        "name": "joe",
      }
    }
  />
  <Items
    items={
      Array [
        Object {
          "id": 1,
          "name": "item 1",
          "price": 7,
        },
        Object {
          "id": 2,
          "name": "item 2",
          "price": 4,
        },
      ]
    }
  />
</Container>
`;

exports[`Items renders one item correctly 1`] = `
<Container>
  <Item
    item={
      Object {
        "id": 1,
        "name": "item 1",
        "price": 7,
      }
    }
    onDelete={[Function]}
  />
  <Total
    total={7}
  />
</Container>
`;

exports[`Items renders two items and deletes correctly 1`] = `
<Container>
  <Item
    item={
      Object {
        "id": 1,
        "name": "item 1",
        "price": 7,
      }
    }
    onDelete={[Function]}
  />
  <Item
    item={
      Object {
        "id": 2,
        "name": "item 2",
        "price": 4,
      }
    }
    onDelete={[Function]}
  />
  <Total
    total={11}
  />
</Container>
`;

exports[`Items renders two items and deletes correctly 2`] = `
<Container>
  <Item
    item={
      Object {
        "id": 2,
        "name": "item 2",
        "price": 4,
      }
    }
    onDelete={[Function]}
  />
  <Total
    total={4}
  />
</Container>
`;

exports[`Items renders zero items correctly 1`] = `
<Container>
  <Total
    total={0}
  />
</Container>
`;

exports[`User renders deep correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
>
  joe - joe@mail.com
</Text>
`;

exports[`User renders shallow correctly 1`] = `
<Title>
  joe - joe@mail.com
</Title>
`;
